packages/grafana-runtime/src/utils/queryResponse.ts

```diff
@@ -24,56 +25,63 @@ interface DataResponse {
 /**
  * Parse the results from /api/ds/query into a DataQueryResponse
  *
+ * @param res - the HTTP response data.
+ * @param queries optional DataQuery array that will order the response based on the order of query refId's.
+ *
  * @public
  */
-export function toDataQueryResponse(res: any): DataQueryResponse {
+export function toDataQueryResponse(res: any, queries?: DataQuery[]): DataQueryResponse {
   const rsp: DataQueryResponse = { data: [], state: LoadingState.Done };
   if (res.data?.results) {
     const results: KeyValue = res.data.results;
-    for (const refId of Object.keys(results)) {
+    const refIDs = queries ? queries.map((q) => q.refId) : Object.keys(results);
+
+    for (const refId of refIDs) {
       const dr = results[refId] as DataResponse;
-      if (dr) {
-        if (dr.error) {
-          if (!rsp.error) {
-            rsp.error = {
-              refId,
-              message: dr.error,
-            };
-            rsp.state = LoadingState.Error;
-          }
+      if (!dr) {
+        continue;
+      }
+
+      if (dr.error) {
+        if (!rsp.error) {
+          rsp.error = {
+            refId,
+            message: dr.error,
+          };
+          rsp.state = LoadingState.Error;
         }
+      }
 
-        if (dr.series && dr.series.length) {
-          for (const s of dr.series) {
-            if (!s.refId) {
-              s.refId = refId;
-            }
-            rsp.data.push(toDataFrame(s));
+      if (dr.series && dr.series.length) {
+        for (const s of dr.series) {
+          if (!s.refId) {
+            s.refId = refId;
           }
+          rsp.data.push(toDataFrame(s));
         }
+      }
 
-        if (dr.tables && dr.tables.length) {
-          for (const s of dr.tables) {
-            if (!s.refId) {
-              s.refId = refId;
-            }
-            rsp.data.push(toDataFrame(s));
+      if (dr.tables && dr.tables.length) {
```

Comment:
```suggestion
      if (dr.tables?.length) {
```
maybe 6 vs 1/2 dozen